// nathcf/index.js
import fetch from 'node-fetch'

const base_url_solve = 'https://cf-solver-renofc.my.id/api/solvebeta'

async function req(payload) {
  const res = await fetch(base_url_solve, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload)
  })

  if (!res.ok) throw new Error(`HTTP ${res.status}`)

  return res.json()
}

function parseProxy(proxy) {
  if (!proxy) return undefined

  const [auth, hostport] = proxy.includes('@')
    ? proxy.split('@')
    : [null, proxy]

  if (!hostport || !hostport.includes(':')) return undefined

  const [host, port] = hostport.split(':')
  if (!host || !port || isNaN(port)) return undefined

  if (auth) {
    const [username, password] = auth.split(':')
    return { host, port: Number(port), username, password }
  }

  return { host, port: Number(port) }
}

export const nathcf = {

  async wafSession(url, proxy) {
    if (!url) throw new Error('url wajib')

    const data = await req({
      url,
      mode: 'waf-session',
      proxy: parseProxy(proxy)
    })

    return {
      cookies: data.cookies ?? null,
      headers: data.headers ?? null,
      cf_clearance: data.cf_clearance ?? null
    }
  },

  async turnstileMin(url, sitekey, proxy) {
    if (!url || !sitekey) throw new Error('url dan sitekey wajib')

    const data = await req({
      url,
      sitekey,
      mode: 'turnstile-min',
      proxy: parseProxy(proxy)
    })

    return {
      token: data?.result?.token ?? null,
      time: data?.time ?? null
    }
  },

  async turnstileMax(url, proxy) {
    if (!url) throw new Error('url wajib')

    const data = await req({
      url,
      mode: 'turnstile-max',
      proxy: parseProxy(proxy)
    })

    return {
      token: data?.result?.token ?? null,
      time: data?.time ?? null
    }
  },

  async recaptchav3(url, sitekey, proxy) {
    if (!url || !sitekey) throw new Error('url dan sitekey wajib')

    const data = await req({
      url,
      sitekey,
      mode: 'recaptcha-min',
      proxy: parseProxy(proxy)
    })

    return {
      token: data?.result?.recaptchaToken ?? null,
      time: data?.time ?? null
    }
  },

  async source(url, proxy) {
    if (!url) throw new Error('url wajib')

    const data = await req({
      url,
      mode: 'source',
      proxy: parseProxy(proxy)
    })

    return typeof data === 'string' ? data : data?.html ?? data
  },

  async key(url, proxy) {
    if (!url) throw new Error('url wajib')

    const data = await req({
      url,
      mode: 'key',
      proxy: parseProxy(proxy)
    })

    return {
      sitekey: data?.data?.sitekey ?? null,
      type: data?.type ?? null,
      source: data?.source ?? null,
      time: data?.time ?? null
    }
  }

}

export default { nathcf }
